pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs

```diff
@@ -0,0 +1,190 @@
+            keys: get_dispatch.keys,
+            height: u256_to_u64(get_dispatch.height)?,
+            timeout_timestamp: u256_to_u64(get_dispatch.timeoutTimestamp)?,
+        };
+        handle.record_cost(cost)?;
+        match dispatcher.dispatch_request(DispatchRequest::Get(get_dispatch)) {
+            Ok(_) => {
+                let nonce = Pallet::<T>::previous_nonce();
+                GasLimits::<T>::insert(nonce, gas_limit);
+                Ok(PrecompileOutput { exit_status: ExitSucceed::Stopped, output: vec![] })
+            }
+            Err(e) => Err(PrecompileFailure::Error {
+                exit_status: ExitError::Other(format!("dispatch execution failed: {:?}", e).into()),
+            }),
+        }
+    }
+}
+
+/// Ismp Response Dispatcher precompile for evm contracts
+pub struct IsmpResponseDispatcher<T> {
+    _marker: PhantomData<T>,
+}
+
+impl<T> Precompile for IsmpResponseDispatcher<T>
+where
+    T: Config + pallet_evm::Config,
+    <T as frame_system::Config>::Hash: From<H256>,
+{
+    fn execute(handle: &mut impl PrecompileHandle) -> PrecompileResult {
+        let input = handle.input();
+
+        let weight = <T as Config>::WeightInfo::dispatch_response();
+
+        let cost = <T as pallet_evm::Config>::GasWeightMapping::weight_to_gas(weight);
+
+        let dispatcher = Dispatcher::<T>::default();
+        let post_response =
+            SolPostResponse::decode(input, true).map_err(|e| PrecompileFailure::Error {
+                exit_status: ExitError::Other(format!("Failed to decode input: {:?}", e).into()),
+            })?;
+        let post_response = PostResponse {
+            post: Post {
+                source: parse_state_machine(post_response.request.source)?,
+                dest: parse_state_machine(post_response.request.dest)?,
+                nonce: u256_to_u64(post_response.request.nonce)?,
+                from: post_response.request.from,
+                to: post_response.request.to,
+                timeout_timestamp: u256_to_u64(post_response.request.timeoutTimestamp)?,
+                data: post_response.request.data,
+            },
+            response: post_response.response,
+        };
+        handle.record_cost(cost)?;
+
+        match dispatcher.dispatch_response(post_response) {
+            Ok(_) => Ok(PrecompileOutput { exit_status: ExitSucceed::Stopped, output: vec![] }),
+            Err(e) => Err(PrecompileFailure::Error {
+                exit_status: ExitError::Other(format!("dispatch execution failed: {:?}", e).into()),
+            }),
+        }
+    }
+}
+
+/// Convert u256 to u64 with overflow check
+fn u256_to_u64(value: alloy_primitives::U256) -> Result<u64, PrecompileFailure> {
+    let value = U256::from_big_endian(value.to_be_bytes::<32>().as_slice());
+    for i in &value.0[1..] {
+        if *i != 0u64 {
+            return Err(PrecompileFailure::Error {
+                exit_status: ExitError::Other("Integer Overflow".into()),
+            })
+        }
```

Comment:
why would it overflow?

Reply:
it should not ideally, I'll add it as a comment instead and use the default method.